File: 프로그래머스/lv2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.cpp
#include <string>
#include <vector>
#include<iostream>
#include<map>
#include<algorithm>
using namespace std;

map<string,int>m;

bool cmp(pair<int,string> a,pair<int,string> b){
    return a.first>b.first; // 많이 주문된 순서
}

void count(vector<char>v){
    
    sort(v.begin(),v.end()); // 메뉴 구성을 알파벳 순서로 정렬 
    string str="";
    for(auto a:v){
        str+=a;
    }
    m[str]++; // map에 해당하는 메뉴가 몇번 나왔는지 저장 
    return;
}

void combi(int start,vector<char>v,int size,string str){
    if(v.size()==size){ // 내가 원하는 코스 종류와 같아지면 
        count(v);
        return;
    }
    
    for(int i=start+1;i<str.size();i++){ // 조합을 통해서 n개의 모든 메뉴 조합 보겠다 
        v.push_back(str[i]); 
        combi(i,v,size,str);
        v.pop_back();
    }
    return;
    
}

vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    
    for(int k=0;k<course.size();k++){ // 원하는 코스 요리 개수 정하고 
        m.clear();
        for(int i=0;i<orders.size();i++){ // 모든 사람의 메뉴 구성을 본다 
            vector<char>v;
            combi(-1,v,course[k],orders[i]);
        }
        
        vector<pair<int,string>>menu;
        
        for(auto a:m){
            if(a.second>1) // 2명 이상에게 주문되야 
                menu.push_back({a.second,a.first});
        }
        
        sort(menu.begin(),menu.end(),cmp); // 가장 많이 주문된 순으로 정렬 
        
        int max=menu[0].first;
        for(int i=0;i<menu.size();i++){ 
            if(max==menu[i].first)  // 동일한 최고 횟수로 주문된 메뉴는 전부 넣어준다 
                answer.push_back(menu[i].second);
            else
                break;
        }
    }
    
    
    sort(answer.begin(),answer.end()); // 최종적으로 오름차순 정렬
    
    
    return answer;
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;
import java.util.stream.*;
class Solution {
        
    private static final List<Pair>list= new ArrayList<>();
    
    private static int MAX = 0;
    
    public String[] solution(String[] orders, int[] course) {

        for(int length : course){
            combis(orders,length);
        }
        
        
        list.sort((p1,p2)->p1.key.compareTo(p2.key));
        
        String []answer = new String[list.size()];
        
        for(int i=0;i<list.size();i++){
            answer[i] = list.get(i).key;
        }
        
        return answer;
    }
    
    private void combis(String orders[],int length){
        
        Map<String,Pair>m = new HashMap<>();
        MAX = 0;
        
        for(String order : orders){
            if(order.length()<length)
                continue;
            combi(m,order,length,new StringBuilder(),-1);
        }
        
        list.addAll(m.values()
            .stream()
            .filter(p->p.count==MAX&&p.count>=2)
            .sorted((p1,p2)->{
                return p1.key.compareTo(p2.key);
            })
            .collect(Collectors.toList())
                    );
        
    }
    
    private void combi(Map<String,Pair>m,String order,int length,StringBuilder sb,int depth){
        
        if(length==sb.length()){
            String str = sb
                .toString()
                .chars()
                .boxed()
                .sorted((v1,v2)->v1-v2)
                .collect(
                StringBuilder::new,
                StringBuilder::appendCodePoint,
                StringBuilder::append
            ).toString();
                        
            if(m.containsKey(str)){
                Pair p = m.get(str);
                p.add();
                MAX=Math.max(MAX,p.count);
            }else{
                m.put(str,new Pair(str,1));
                MAX=Math.max(MAX,1);
            }
            return;
        }        
        
        for(int i = depth+1;i<order.length();i++){
            sb.append(order.charAt(i));
            combi(m,order,length,sb,i);
            sb.deleteCharAt(sb.length()-1);
        }
    }
    
}

class Pair{
    
    String key;
    int count;
    
    public Pair(String key,int count){
        this.key=key;
        this.count=count;
    }
    
    public void add(){
        this.count++;
    }
    
}